entity add issue when generate light boom

```shell
Caused by: java.lang.IllegalArgumentException: Entity is already registered com.badlogic.ashley.core.Entity@285c08c8
	at com.badlogic.ashley.core.EntityManager.addEntityInternal(EntityManager.java:117)
	at com.badlogic.ashley.core.EntityManager.addEntity(EntityManager.java:33)
	at com.badlogic.ashley.core.Engine.addEntity(Engine.java:82)
	at com.voidvvv.game.skill.CastLightBoom.createLightBoom(CastLightBoom.java:79)
	... 10 more

```
it occurs when you use CastLightBoom in a very high frequency.

it seems like the entity is not removed properly from engine when reset actor.
need to investigate the root cause

--- a/core/src/main/java/com/voidvvv/game/actor/Bob.java
+++ b/core/src/main/java/com/voidvvv/game/actor/Bob.java
@@ -47,8 +47,10 @@ public class Bob extends MoveShapeBox2dActor {
     public void init() {
         super.init();
         StateMachine machine = new DefaultStateMachine(this, new Idle());
-        this.getEntity().add(new StateMachineComponent(machine));
-        this.getEntity().add(AssetUtils.cpy(animPrototype));
+        StateMachineComponent stateMachineComponent = Pools.obtain(StateMachineComponent.class);
+        stateMachineComponent.setStateMachine(machine);
+        this.getEntity().add(stateMachineComponent);
+        this.getEntity().add(animPrototype);
 
         MoveChangeListenerComponent moveChangeListenerComponent = getEntity().getComponent(MoveChangeListenerComponent.class);
         moveChangeListenerComponent.list.add( () -> {

--- a/core/src/main/java/com/voidvvv/game/actor/MoveShapeBox2dActor.java
+++ b/core/src/main/java/com/voidvvv/game/actor/MoveShapeBox2dActor.java
@@ -5,6 +5,7 @@ import com.badlogic.ashley.core.Entity;
 import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.physics.box2d.Fixture;
 import com.badlogic.gdx.physics.box2d.World;
+import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.Main;
 import com.voidvvv.game.base.VRectBoundComponent;
 import com.voidvvv.game.battle.BattleContext;
@@ -29,13 +30,13 @@ public class MoveShapeBox2dActor extends VFlatWorldActor {
     public void init() {
         super.init();
         this.setDead(false);
-        this.getEntity().add(new MoveComponent());
-        this.getEntity().add(new VBox2dComponent());
-        this.getEntity().add(new VRectBoundComponent());
-        this.getEntity().add(new BattleEventListenerComponent());
-        this.getEntity().add(new DefaultBattleComponent());
-        this.getEntity().add(new MoveChangeListenerComponent());
-        this.getEntity().add(new ContactTypeComponent());
+        this.getEntity().add(Pools.obtain(MoveComponent.class));
+        this.getEntity().add(Pools.obtain(VBox2dComponent.class));
+        this.getEntity().add(Pools.obtain(VRectBoundComponent.class));
+        this.getEntity().add(Pools.obtain(BattleEventListenerComponent.class));
+        this.getEntity().add(Pools.obtain(DefaultBattleComponent.class));
+        this.getEntity().add(Pools.obtain(MoveChangeListenerComponent.class));
+        this.getEntity().add(Pools.obtain(ContactTypeComponent.class));
 
     }
 
@@ -67,19 +68,6 @@ public class MoveShapeBox2dActor extends VFlatWorldActor {
     public void unload() {
         super.unload();
 
-        VBox2dComponent box2dComponent = this.entity.getComponent(VBox2dComponent.class);
-        if (box2dComponent != null) {
-            World world = box2dComponent.getFlatBody().getWorld();
-            world.destroyBody(box2dComponent.getFlatBody());
-            box2dComponent.setBottomFixture(null);
-            box2dComponent.setFaceFixture(null);
-            box2dComponent.setFlatBody(null);
-//            box2dComponent.clearEndContactFixtures();
-//            box2dComponent.clearStartContactFixtures();
-            box2dComponent.clearContactPairListener();
-        }
-        MoveChangeListenerComponent moveChangeListenerComponent = getEntity().getComponent(MoveChangeListenerComponent.class);
-        moveChangeListenerComponent.list.clear();
 
     }
 }

--- a/core/src/main/java/com/voidvvv/game/actor/Slime.java
+++ b/core/src/main/java/com/voidvvv/game/actor/Slime.java
@@ -80,7 +80,7 @@ public class Slime extends MoveShapeBox2dActor {
         this.getEntity().getComponent(MoveComponent.class).speed = 100f;
         getEntity().getComponent(VBox2dComponent.class).addContactPairListener(this::contact);
 
-        this.getEntity().add(AssetUtils.cpy(animPrototype));
+        this.getEntity().add(animPrototype);
         StateMachine machine = new DefaultStateMachine(this, new Idle());
         this.getEntity().add(new StateMachineComponent(machine));
 

--- a/core/src/main/java/com/voidvvv/game/actor/bulltes/BaseBullet.java
+++ b/core/src/main/java/com/voidvvv/game/actor/bulltes/BaseBullet.java
@@ -3,6 +3,7 @@ package com.voidvvv.game.actor.bulltes;
 import com.badlogic.ashley.core.Entity;
 import com.badlogic.gdx.math.Vector2;
 import com.badlogic.gdx.physics.box2d.World;
+import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.base.VRectBoundComponent;
 import com.voidvvv.game.battle.DefaultBattleComponent;
 import com.voidvvv.game.box2d.CollisionPair;
@@ -28,10 +29,12 @@ public class BaseBullet extends VFlatWorldActor {
     @Override
     public void init() {
         super.init();
-        this.getEntity().add(new MoveComponent());
-        this.getEntity().add(new VBox2dComponent());
-        this.getEntity().add(new VRectBoundComponent());
-        this.getEntity().add(new ContactTypeComponent(ContactTypeComponent.BULLET));
+        this.getEntity().add(Pools.obtain(MoveComponent.class));
+        this.getEntity().add(Pools.obtain(VBox2dComponent.class));
+        this.getEntity().add(Pools.obtain(VRectBoundComponent.class));
+        ContactTypeComponent obtain = Pools.obtain(ContactTypeComponent.class);
+        obtain.type = ContactTypeComponent.BULLET;
+        this.getEntity().add(obtain);
 
         this.getEntity().getComponent(MoveComponent.class).speed = initSpeed;
         getEntity().getComponent(VBox2dComponent.class).addContactPairListener(this::contact);
@@ -60,17 +63,6 @@ public class BaseBullet extends VFlatWorldActor {
     public void unload() {
         super.unload();
         owner = null;
-        VBox2dComponent box2dComponent = this.entity.getComponent(VBox2dComponent.class);
-        if (box2dComponent != null) {
-            World world = box2dComponent.getFlatBody().getWorld();
-            world.destroyBody(box2dComponent.getFlatBody());
-            box2dComponent.setBottomFixture(null);
-            box2dComponent.setFaceFixture(null);
-            box2dComponent.setFlatBody(null);
-//            box2dComponent.clearEndContactFixtures();
-//            box2dComponent.clearStartContactFixtures();
-            box2dComponent.clearContactPairListener();
-        }
     }
 
 }

--- a/core/src/main/java/com/voidvvv/game/actor/bulltes/LightBoom.java
+++ b/core/src/main/java/com/voidvvv/game/actor/bulltes/LightBoom.java
@@ -6,6 +6,7 @@ import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.graphics.Texture;
 import com.badlogic.gdx.graphics.g2d.Animation;
 import com.badlogic.gdx.graphics.g2d.TextureRegion;
+import com.badlogic.gdx.math.MathUtils;
 import com.badlogic.gdx.physics.box2d.Fixture;
 import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.Main;
@@ -24,21 +25,32 @@ import com.voidvvv.game.utils.AssetConstants;
 import com.voidvvv.game.utils.ReflectUtil;
 import com.voidvvv.render.actor.VActorRender;
 
+import java.security.SecureRandom;
+import java.util.Random;
+import java.util.random.RandomGenerator;
+
 public class LightBoom extends BaseBullet {
     public static final ComponentMapper<BattleContextComponent> battleContextComponentMapper = ComponentMapper.getFor(BattleContextComponent.class);
 
     public static SimpleAnimateComponent simpleAnimateComponent;
 
+    public long randomNum = 0L;
     public LightBoom() {
         if (simpleAnimateComponent == null) {
             initAnimate();
         }
         Gdx.app.log("LightBoom", "LightBoom constructor");
+        this.randomNum = RandomGenerator.getDefault().nextLong();
     }
 
     public static LightBoom create() {
         Gdx.app.log("LightBoom", "LightBoom create");
         LightBoom obtain = Pools.obtain(LightBoom.class);
+        boolean contains = Main.getInstance().getGameMode().getEngine().getEntities().contains(obtain.getEntity(), true);
+        if (contains) {
+            Gdx.app.log("LightBoom", "error detect! " + obtain.randomNum);
+        }
+        Gdx.app.log("LightBoom", "LightBoom create " + obtain.randomNum);
         return obtain;
     }
 
@@ -112,4 +124,10 @@ public class LightBoom extends BaseBullet {
 
     }
 
+
+    @Override
+    public void reset() {
+        super.reset();
+        Gdx.app.log("LightBoom", "LightBoom reset " + this.randomNum);
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/base/VRectBoundComponent.java
+++ b/core/src/main/java/com/voidvvv/game/base/VRectBoundComponent.java
@@ -2,8 +2,9 @@ package com.voidvvv.game.base;
 
 import com.badlogic.ashley.core.Component;
 import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.Pool;
 
-public class VRectBoundComponent extends VPositionComponent implements Component {
+public class VRectBoundComponent extends VPositionComponent implements Component, Pool.Poolable {
     private float width; // z len
     private float height; // y len
     private float length; // x len
@@ -68,4 +69,11 @@ public class VRectBoundComponent extends VPositionComponent implements Component
         length = 0;
     }
 
+    @Override
+    public void reset() {
+        width = 0;
+        height = 0;
+        length = 0;
+        position.set(0, 0);
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/base/world/VWorldActor.java
+++ b/core/src/main/java/com/voidvvv/game/base/world/VWorldActor.java
@@ -4,7 +4,9 @@ package com.voidvvv.game.base.world;
 import com.badlogic.ashley.core.Component;
 import com.badlogic.ashley.core.Engine;
 import com.badlogic.ashley.core.Entity;
+import com.badlogic.ashley.utils.ImmutableArray;
 import com.badlogic.gdx.ai.fsm.DefaultStateMachine;
+import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.Main;
 import com.voidvvv.game.base.StateComponentHolder;
 import com.voidvvv.game.base.VActor;
@@ -65,9 +67,10 @@ public abstract class VWorldActor implements VActor {
     public void dispose() {
         VActor.super.dispose();
     }
-
+    public boolean ready = true;
     @Override
     public void reset() {
+        ready = true;
         unload();
         Engine engine = Main.getInstance().getGameMode().getEngine();
         engine.removeEntity(entity);
@@ -75,5 +78,10 @@ public abstract class VWorldActor implements VActor {
     }
 
     protected void unload() {
+        ImmutableArray<Component> components = entity.getComponents();
+        for (int i = 0; i < components.size(); i++) {
+            Component component = components.get(i);
+            Pools.free(component);
+        }
     }
 }

--- a/core/src/main/java/com/voidvvv/game/box2d/VBox2dComponent.java
+++ b/core/src/main/java/com/voidvvv/game/box2d/VBox2dComponent.java
@@ -3,12 +3,14 @@ package com.voidvvv.game.box2d;
 import com.badlogic.ashley.core.Component;
 import com.badlogic.gdx.physics.box2d.Body;
 import com.badlogic.gdx.physics.box2d.Fixture;
+import com.badlogic.gdx.physics.box2d.World;
+import com.badlogic.gdx.utils.Pool;
 
 import java.security.PublicKey;
 import java.util.ArrayList;
 import java.util.List;
 
-public class VBox2dComponent implements Component {
+public class VBox2dComponent implements Component, Pool.Poolable {
 
     Body flatBody;
 
@@ -105,4 +107,15 @@ public class VBox2dComponent implements Component {
     }
 
 
+    @Override
+    public void reset() {
+        World world = this.getFlatBody().getWorld();
+        world.destroyBody(this.getFlatBody());
+        this.setBottomFixture(null);
+        this.setFaceFixture(null);
+        this.setFlatBody(null);
+//            box2dComponent.clearEndContactFixtures();
+//            box2dComponent.clearStartContactFixtures();
+        this.clearContactPairListener();
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/BattleContextComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/BattleContextComponent.java
@@ -10,6 +10,13 @@ public class BattleContextComponent implements Component {
         this.battleContext = battleContext;
     }
 
+    public void setBattleContext(BattleContext battleContext) {
+        this.battleContext = battleContext;
+    }
+
+    public BattleContextComponent() {
+    }
+
     public BattleContext getBattleContext() {
         return battleContext;
     }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/BattleEventListenerComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/BattleEventListenerComponent.java
@@ -1,12 +1,18 @@
 package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
+import com.badlogic.gdx.utils.Pool;
 import com.voidvvv.game.battle.BattleEventListener;
 
 import java.util.ArrayList;
 import java.util.List;
 
-public class BattleEventListenerComponent implements Component {
+public class BattleEventListenerComponent implements Component, Pool.Poolable {
+    @Override
+    public void reset() {
+        listeners.clear();
+    }
+
     List<BattleEventListener> listeners = new ArrayList<>();
 
     public void addListener(BattleEventListener listener) {

--- a/core/src/main/java/com/voidvvv/game/ecs/components/CampComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/CampComponent.java
@@ -6,6 +6,9 @@ import com.voidvvv.game.camp.CampContext;
 public class CampComponent implements Component {
     private long campSign;
 
+    public CampComponent() {
+    }
+
     public CampComponent(long campSign) {
         this.campSign = campSign;
     }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/ContactTypeComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/ContactTypeComponent.java
@@ -1,8 +1,9 @@
 package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
+import com.badlogic.gdx.utils.Pool;
 
-public class ContactTypeComponent implements Component {
+public class ContactTypeComponent implements Component, Pool.Poolable {
     public static final int CREATURE = 0;
     public static final int WALL = 1;
     public static final int BULLET = 2;
@@ -14,5 +15,10 @@ public class ContactTypeComponent implements Component {
 
     public ContactTypeComponent() {
         this(CREATURE);
+    }
+
+    @Override
+    public void reset() {
+        type = CREATURE;
     }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/MoveChangeListenerComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/MoveChangeListenerComponent.java
@@ -1,11 +1,17 @@
 package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
+import com.badlogic.gdx.utils.Pool;
 import com.voidvvv.game.base.MoveChangeListener;
 
 import java.util.ArrayList;
 import java.util.List;
 
-public class MoveChangeListenerComponent implements Component {
+public class MoveChangeListenerComponent implements Component, Pool.Poolable {
     public final List<MoveChangeListener> list = new ArrayList<>();
+
+    @Override
+    public void reset() {
+        list.clear();
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/MoveComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/MoveComponent.java
@@ -2,9 +2,10 @@ package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
 import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.Pool;
 import com.voidvvv.game.base.components.VComponent;
 
-public class MoveComponent implements VComponent, Component {
+public class MoveComponent implements VComponent, Component, Pool.Poolable {
     public final Vector2 vel = new Vector2();
     public final Vector2 preVel = new Vector2();
     public float speed;
@@ -29,5 +30,10 @@ public class MoveComponent implements VComponent, Component {
     @Override
     public void dispose() {
 
+    }
+
+    @Override
+    public void reset() {
+        init();
     }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/StateMachineComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/StateMachineComponent.java
@@ -2,12 +2,16 @@ package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
 import com.badlogic.gdx.ai.fsm.StateMachine;
+import com.badlogic.gdx.utils.Pool;
 import com.voidvvv.game.base.components.VComponent;
 
-public class StateMachineComponent implements Component {
+public class StateMachineComponent implements Component, Pool.Poolable {
     StateMachine stateMachine;
 
     public float stateTime = 0f;
+
+    public StateMachineComponent() {
+    }
 
     public StateMachineComponent(StateMachine stateMachine) {
         this.stateMachine = stateMachine;
@@ -21,4 +25,9 @@ public class StateMachineComponent implements Component {
         this.stateMachine = stateMachine;
     }
 
+    @Override
+    public void reset() {
+        stateMachine = null;
+        stateTime = 0f;
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/components/TimeComponent.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/components/TimeComponent.java
@@ -1,8 +1,13 @@
 package com.voidvvv.game.ecs.components;
 
 import com.badlogic.ashley.core.Component;
+import com.badlogic.gdx.utils.Pool;
 
-public class TimeComponent implements Component {
+public class TimeComponent implements Component, Pool.Poolable {
     public float time;
 
+    @Override
+    public void reset() {
+        time = 0;
+    }
 }

--- a/core/src/main/java/com/voidvvv/game/ecs/system/VWorldActorManageSystem.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/system/VWorldActorManageSystem.java
@@ -3,6 +3,7 @@ package com.voidvvv.game.ecs.system;
 import com.badlogic.ashley.core.Entity;
 import com.badlogic.ashley.core.Family;
 import com.badlogic.ashley.systems.IteratingSystem;
+import com.badlogic.gdx.Gdx;
 import com.voidvvv.game.base.VActor;
 import com.voidvvv.game.base.VRectBoundComponent;
 import com.voidvvv.game.base.world.VWorldActor;
@@ -26,6 +27,7 @@ public class VWorldActorManageSystem extends IteratingSystem {
     }
 
     private void flushActors(VWorldActorComponent component) {
+        Gdx.app.log("VWorldActorManageSystem", "============= flushActors start ==========");
         for (VWorldActor actor : component.toRemove) {
             actor.dispose();
             component.actors.remove(actor);
@@ -37,6 +39,8 @@ public class VWorldActorManageSystem extends IteratingSystem {
         }
         component.toAdd.clear();
         component.actors.sort(VActorComparator.INSTANCE);
+        Gdx.app.log("VWorldActorManageSystem", "============= flushActors end ==========");
+
     }
 
     public static class VActorComparator implements Comparator<VActor> {

--- a/core/src/main/java/com/voidvvv/game/ecs/system/debug/ConstantCastSkill.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/system/debug/ConstantCastSkill.java
@@ -1,0 +1,24 @@
+package com.voidvvv.game.ecs.system.debug;
+
+import com.badlogic.ashley.core.Entity;
+import com.badlogic.ashley.core.Family;
+import com.badlogic.ashley.systems.IntervalIteratingSystem;
+import com.badlogic.ashley.systems.IteratingSystem;
+import com.voidvvv.game.ecs.components.skill.MainSkillComponent;
+
+public class ConstantCastSkill extends IntervalIteratingSystem {
+
+
+    public ConstantCastSkill() {
+        super(Family.all(MainSkillComponent.class).get(), 0.1f);
+        // super();
+    }
+
+    @Override
+    protected void processEntity(Entity entity) {
+        MainSkillComponent component = entity.getComponent(MainSkillComponent.class);
+        if (component != null) {
+            component.skill.cast();
+        }
+    }
+}

--- a/core/src/main/java/com/voidvvv/game/ecs/system/simple/SimpleSlimeGenerateStrategy.java
+++ b/core/src/main/java/com/voidvvv/game/ecs/system/simple/SimpleSlimeGenerateStrategy.java
@@ -9,6 +9,7 @@ import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.math.MathUtils;
 import com.badlogic.gdx.math.Rectangle;
 import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.actor.ActorConstants;
 import com.voidvvv.game.actor.Slime;
 import com.voidvvv.game.actor.utils.ActorMetaData;
@@ -76,8 +77,9 @@ public class SimpleSlimeGenerateStrategy extends IntervalSystem {
             }
 
         }
-
-        slime.getEntity().add(new CampComponent(CampConstants.BLACK));
+        CampComponent obtain = Pools.obtain(CampComponent.class);
+        obtain.setCampSign(CampConstants.BLACK);
+        slime.getEntity().add(obtain);
         slime.getEntity().getComponent(BattleEventListenerComponent.class)
             .addListener(new BattleEventListener() {
                 @Override
@@ -94,7 +96,7 @@ public class SimpleSlimeGenerateStrategy extends IntervalSystem {
                     }
                 }
             });
-        slime.getEntity().add(new EnemySignComponent());
+        slime.getEntity().add(Pools.obtain(EnemySignComponent.class));
         return slime;
     }
 

--- a/core/src/main/java/com/voidvvv/game/impl/flat/VFlatWorld.java
+++ b/core/src/main/java/com/voidvvv/game/impl/flat/VFlatWorld.java
@@ -220,7 +220,10 @@ public class VFlatWorld implements VWorld {
 
     @Override
     public void resetVActor(VWorldActor actor) {
-        this.actorComponent.resetActor(actor);
+        if (actor.ready) {
+            actor.ready = false;
+            this.actorComponent.resetActor(actor);
+        }
     }
 
     @Override

--- a/core/src/main/java/com/voidvvv/game/impl/flat/VFlatWorldActor.java
+++ b/core/src/main/java/com/voidvvv/game/impl/flat/VFlatWorldActor.java
@@ -19,14 +19,7 @@ public class VFlatWorldActor extends VWorldActor{
     @Override
     public void init() {
         super.init();
-        if (getEntity().getComponent(VRectBoundComponent.class) != null) {
 
-            getEntity().getComponent(VRectBoundComponent.class).init();
-            getEntity().getComponent(MoveComponent.class).init();
-        }
-        if (getEntity().getComponent(MoveComponent.class) != null) {
-            getEntity().getComponent(MoveComponent.class).init();
-        }
     }
 
     @Override

--- a/core/src/main/java/com/voidvvv/game/mode/impl/SingleFlatWorldMode.java
+++ b/core/src/main/java/com/voidvvv/game/mode/impl/SingleFlatWorldMode.java
@@ -9,6 +9,7 @@ import com.badlogic.gdx.ai.msg.MessageManager;
 import com.badlogic.gdx.ai.msg.Telegram;
 import com.badlogic.gdx.graphics.g2d.SpriteBatch;
 import com.badlogic.gdx.math.Vector2;
+import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.Main;
 import com.voidvvv.game.actor.ActorConstants;
 import com.voidvvv.game.actor.Alice;
@@ -28,6 +29,7 @@ import com.voidvvv.game.ecs.components.*;
 import com.voidvvv.game.base.world.VActorSpawnHelper;
 import com.voidvvv.game.base.world.WorldContext;
 import com.voidvvv.game.ecs.system.*;
+import com.voidvvv.game.ecs.system.debug.ConstantCastSkill;
 import com.voidvvv.game.ecs.system.render.DebugRenderIteratorSystem;
 import com.voidvvv.game.ecs.system.render.EntityRenderSystem;
 import com.voidvvv.game.ecs.system.simple.SimpleSlimeGenerateStrategy;
@@ -151,7 +153,9 @@ public class SingleFlatWorldMode implements VWorldContextGameMode, TimeLimitMode
 
             }
         });
-        entity.add(new BattleContextComponent(baseBattleContext));
+        BattleContextComponent obtain = Pools.obtain(BattleContextComponent.class);
+        obtain.setBattleContext(baseBattleContext);
+        entity.add(obtain);
         entity.add(damageValueComponent);
         entity.add(new CampContextComponent(new CampContext()));
 
@@ -171,6 +175,9 @@ public class SingleFlatWorldMode implements VWorldContextGameMode, TimeLimitMode
 
         // autogenerate slime
         engine.addSystem(new SimpleSlimeGenerateStrategy(context));
+
+        // for debug
+        engine.addSystem(new ConstantCastSkill());
 
         moveMapper = ComponentMapper.getFor(MoveComponent.class);
     }

--- a/core/src/main/java/com/voidvvv/game/skill/CastLightBoom.java
+++ b/core/src/main/java/com/voidvvv/game/skill/CastLightBoom.java
@@ -1,6 +1,7 @@
 package com.voidvvv.game.skill;
 
 import com.badlogic.ashley.core.Entity;
+import com.badlogic.gdx.math.MathUtils;
 import com.badlogic.gdx.math.Vector2;
 import com.badlogic.gdx.utils.Pools;
 import com.voidvvv.game.Main;
@@ -47,13 +48,22 @@ public class CastLightBoom implements Skill {
 
     @Override
     public void cast() {
-        LightBoom lightBoom = createLightBoom();
-        lightBoom.getEntity().getComponent(MoveComponent.class).vel.set(direction);
+        for (int x = 0; x < 8; x++) {
+            float percent = x / 8f;
+            float angle = percent * MathUtils.PI*2;
+            direction.set(MathUtils.cos(angle), MathUtils.sin(angle));
+            LightBoom lightBoom = createLightBoom();
+            lightBoom.getEntity().getComponent(MoveComponent.class).vel.set(direction);
+        }
     }
+
     Vector2 direction = new Vector2(1, 0);
     float speed = 300f;
-    public LightBoom createLightBoom () {
+
+    public LightBoom createLightBoom() {
         LightBoom lightBoom = LightBoom.create();
+
+
         VActorSpawnHelper helper = Pools.obtain(VActorSpawnHelper.class);
         helper.hx = META_DATE.getRectProps().getLength() / 2f;
         helper.hy = META_DATE.getRectProps().getHeight() / 2f;
@@ -67,7 +77,12 @@ public class CastLightBoom implements Skill {
         }
         lightBoom.setWorldContext(this.getWorldContext());
         lightBoom.owner = this.getOwner();
-        Main.getInstance().getGameMode().getEngine().addEntity(lightBoom.getEntity());
+        try {
+            Main.getInstance().getGameMode().getEngine().addEntity(lightBoom.getEntity());
+        } catch (Exception e) {
+            System.out.println(e);
+            throw new RuntimeException(e);
+        }
         worldContext.getWorld().spawnVActor(() -> {
             return lightBoom;
         }, helper);

